src/master_client.h

```diff
@@ -46,10 +46,10 @@ class MasterClient {
   // Properties of various watermarks (and their extreme cases):
   //   sn_ckpt <= sn_latest_tail + 1 (i.e., everything has been checkpointed)
   //   sn_flushed <= sn_ckpt (i.e., all checkpointed data has been flushed)
-  Status GetWatermark(Watermark w, int64_t* val) const;
-  Status SetWatermark(Watermark w, int64_t new_val);
+  virtual Status GetWatermark(Watermark w, int64_t* val) const = 0;
+  virtual Status SetWatermark(Watermark w, int64_t new_val) = 0;
 
- private:
+ protected:
```

Comment:
Relevant rule from the style guide: "Limit the use of protected to those member functions that might need to be accessed from subclasses. Note that data members should be private." http://google.github.io/styleguide/cppguide.html#Inheritance

Regardless, `redis_context_` can definitely be removed and only stored by `RedisMasterClient`.